### PR TITLE
iGemm: Reduce address calculation for stride > 1

### PR DESCRIPTION
**Pull request by zjing14**
- Reduce addr calculation for stride > 1
- Can achieve 30\~40% improvement for 1x1 convolution fp32, 10\~15% for fp16
- Support fwd/wrw, fp32/fp16, group and non-group

### CODE REVIEW

**Comment by TejashShah** [comment]
@zjing14 Does it also require buffer load?

**Comment by zjing14** [comment]
No. fp16 also use it.

**Comment by zjing14** [comment]
@TejashShah I am testing the improvement on fp16.

**Comment by zjing14** [comment]
MI100-A0 node:
`./bin/MIOpenDriver convfp16 -F 1 -n 64 -c 256 -H 28 -W 28 -k 256 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -t 1`
Develop (tuned): 0.102596 ms; this PR(tuned): 0.090524 ms; Speedup: 1.13

`./bin/MIOpenDriver convfp16 -F 1 -n 64 -c 256 -H 56 -W 56 -k 256 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -t 1`
Develop (tuned): 0.311538 ms; this PR(tuned): 0.270276 ms; Speedup: 1.15

File: src/kernels/composable_kernel/include/tensor_operation/threadwise_generic_tensor_slice_copy.hpp [hunk 1]
```diff
@@ -110,14 +111,32 @@ struct ThreadwiseGenericTensorSliceCopy_v4r2
...
-                                  SrcDataPerRead,
-                                  SrcAddressSpace,
-                                  AddressSpace::Vgpr,
-                                  InMemoryDataOperation::Set>(
-                        p_src, src_coord.GetOffset(), p_src_long_vector, buffer_offset);
+                    static_if<SrcDataStride == 2>{}([&](auto) {
+                        const auto src_offset = src_coord.GetOffset();
+                        for(index_t j = 0; j < SrcDataPerRead; j++)
+                        {
+                            transfer_data<SrcData,
```
[the remark this answers is not in the document]
Comment (asroy) [reply src/kernels/composable_kernel/include/tensor_operation/threadwise_generic_tensor_slice_copy.hpp]: Also, transfer_data should support any value of SrcDataStride

**Comment by asroy** [comment]
Is this PR still WIP?

File: src/kernels/composable_kernel/include/tensor_operation/blockwise_generic_tensor_slice_copy.hpp [hunk 2]
```diff
@@ -31,7 +31,8 @@ template <index_t BlockSize,
           AddressSpace SrcAddressSpace          = AddressSpace::Generic,
           AddressSpace ThreadBufferAddressSpace = AddressSpace::Generic,
           AddressSpace DstAddressSpace          = AddressSpace::Generic,
-          InMemoryDataOperation DstInMemOp      = InMemoryDataOperation::Set>
+          InMemoryDataOperation DstInMemOp      = InMemoryDataOperation::Set,
+          index_t SrcDataStride                 = 1>
```
Comment (asroy) [inline src/kernels/composable_kernel/include/tensor_operation/blockwise_generic_tensor_slice_copy.hpp:35]: Why only SrcDataStride? DstDataStride can also be supported

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp [hunk 3]
```diff
@@ -121,6 +121,7 @@ static inline ConvSolution GetSolutionBase(const ConvolutionContext& ctx,
     const int C              = KernelInputChannelC(ctx);
     const auto hi            = ConvolutionContextInterpreter::GetInputHeightHi(ctx);
     const auto wi            = ConvolutionContextInterpreter::GetInputWidthWi(ctx);
+    const auto wo_           = ConvolutionContextInterpreter::GetOutputWidthWo(ctx);
```
Comment (asroy) [inline src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp:124]: why ```_```?

File: src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_double_buffer.hpp [hunk 4]
```diff
@@ -265,7 +265,8 @@ struct GridwiseConvolutionImplicitGemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_dou
             2,
             GemmBBlockCopySrcDataPerRead_GemmN,
             GemmBBlockCopyDstDataPerWrite_GemmN,
-            CGlobalMemoryDataOperation>{};
+            CGlobalMemoryDataOperation,
+            ConvStrideW>{};
```
Comment (asroy) [inline src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_double_buffer.hpp:269]: why there is stride specified for B but not for A?

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp (hunk 3 above)
Comment (zjing14) [reply src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp:124]: wo already used.

File: src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_double_buffer.hpp (hunk 4 above)
Comment (zjing14) [reply src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_double_buffer.hpp:269]: Stride always happens in B matrix, right?

Comment (zjing14) [reply src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_gnchw_gkcyx_gnkhw_lds_double_buffer.hpp:270]: Fixed.

**Pull request merged**